src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php

```diff
@@ -232,4 +232,39 @@ public function findByTaxon(TaxonInterface $taxon): array
             ->getResult()
         ;
     }
+
+    public function findOneByChannelAndCodeWithAvailableAssociations(ChannelInterface $channel, string $code): ?ProductInterface
+    {
+        $product = $this->createQueryBuilder('o')
+            ->addSelect('association')
+            ->leftJoin('o.associations', 'association')
+            ->innerJoin('association.associatedProducts', 'associatedProduct', 'WITH', 'associatedProduct.enabled = :enabled')
+            ->where('o.code = :code')
+            ->andWhere(':channel MEMBER OF o.channels')
+            ->andWhere('o.enabled = :enabled')
+            ->setParameter('channel', $channel)
+            ->setParameter('code', $code)
+            ->setParameter('enabled', true)
+            ->getQuery()
+            ->getOneOrNullResult()
+        ;
+
+        if (null === $product) {
+            $product = $this->findOneByChannelAndCode($channel, $code);
+
+            $product?->getAssociations()->clear();
```

Comment:
```suggestion
            $product = $this->findOneByChannelAndCode($channel, $code);
            if (null === $product) {
                return null;
            }

            $product->getAssociations()->clear();
```
No point in even trying to apply any additional logic if the product cannot be not found at all.